sru/modules.py

```diff
@@ -633,33 +611,545 @@ def reset_parameters(self):
+        attn_output = attn_output.transpose(0, 1).contiguous().view(tgt_len, bsz, proj_dim)
+
+        attn_output = attn_output * self.alpha + residual
+        if self.normalize_after and self.layer_norm is not None:
+            attn_output = self.layer_norm(attn_output)
+
+        # (tgt_len, bsz, out_dim)
+        attn_output = self.linear3(self.dropout(attn_output))
+        return attn_output
+
+
+class SRUppCell(SRUCell):
+    """
+    A single layer of SRU++, inherited from SRUCell module
+    """
+
+    def forward(self,
+                input: Tensor,
+                c0: Optional[Tensor] = None,
+                mask_pad: Optional[Tensor] = None,
+                attn_mask: Optional[Tensor] = None,
+                memory: Optional[Tensor] = None,
+                memory_mask_pad: Optional[Tensor] = None) -> Tuple[Tensor, Tensor]:
+        """The forward method of the SRU++ layer.
+        """
+
+        if input.dim() != 2 and input.dim() != 3:
+            raise ValueError("Input must be 2 or 3 dimensional")
+
+        batch_size = input.size(-2)
+        if c0 is None:
+            c0 = torch.zeros(batch_size, self.output_size, dtype=input.dtype,
+                             device=input.device)
+
+        # apply layer norm before activation (i.e. before SRU computation)
+        residual = input
+        if self.layer_norm is not None and not self.normalize_after:
+            input = self.layer_norm(input)
+
+        # apply dropout for multiplication
+        if self.training and (self.rnn_dropout > 0):
+            mask = self.get_dropout_mask_((batch_size, input.size(-1)), self.rnn_dropout)
+            input = input * mask.expand_as(input)
+
+        # get the scaling constant; scale_x is a scalar
+        scale_val: Optional[Tensor] = None
+        scale_val = self.scale_x if self.rescale else None
+
+        # get dropout mask
+        mask_c: Optional[Tensor] = None
+        if self.training and (self.dropout > 0):
+            mask_c = self.get_dropout_mask_((batch_size, self.output_size),
+                                            self.dropout)
+
+        # compute U
+        #   U is (length, batch_size, output_size * num_matrices)
+        transform_module = self.transform_module
+        U = transform_module(input, mask_pad=mask_pad,
+                             attn_mask=attn_mask,
+                             memory=memory,
+                             memory_mask_pad=memory_mask_pad)
+        V = self.weight_c
+
+        # apply elementwise recurrence to get hidden states h and c
+        h, c = self.apply_recurrence(U, V,
+                                     residual, c0,
+                                     scale_val,
+                                     mask_c,
+                                     mask_pad)
+
+        if self.layer_norm is not None and self.normalize_after:
+            h = self.layer_norm(h)
+        return h, c
+
+
+class SRUpp(nn.Module):
```

Comment:
how does `SRUpp` relate to `SRU` module? My previous understanding is that SRUpp is SRU where the custom_module includes attention? Are there some additional details that mean that we cannot use the existing `SRU` module?

Reply:
the forward interface changed since we need to support attention related arguments, such as attention mask. the constructor will take additional options too.